File: backend/app/Models/Event.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

#[Fillable([
    'title',
    'description',
    'date',
    'location',
    'organizer_id',
    'status',
    'qr_token',
    'max_capacity',
    'registration_deadline',
])]
class Event extends Model
{
    use HasFactory;

    protected function casts(): array
    {
        return [
            'date' => 'datetime',
            'registration_deadline' => 'datetime',
        ];
    }

    public function organizer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'organizer_id');
    }

    public function participants(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'participations')
            ->withPivot('attended', 'attended_at')
            ->withTimestamps();
    }

    public function participations(): HasMany
    {
        return $this->hasMany(Participation::class);
    }
}

File: mobile/lib/data/models/event.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'user.dart';

part 'event.freezed.dart';
part 'event.g.dart';

@freezed
abstract class Event with _$Event {
  const factory Event({
    required int id,
    required String title,
    required String description,
    required DateTime date,
    required String location,
    @JsonKey(name: 'organizer_id') required int organizerId,
    required String status,
    @JsonKey(name: 'qr_token') String? qrToken,
    @JsonKey(name: 'max_capacity') required int maxCapacity,
    @JsonKey(name: 'registration_deadline') required DateTime registrationDeadline,
    User? organizer,
    List<User>? participants,
  }) = _Event;

  factory Event.fromJson(Map<String, dynamic> json) => _$EventFromJson(json);
}
